htsget-search/src/htsget/cram_search.rs

```diff
@@ -0,0 +1,324 @@
+        &self,
+        crai_index: &[crai::Record],
+        cram_reader: &mut Reader<BufReader<File>>,
+    ) -> Result<Vec<BytesRange>> {
+        Self::get_bytes_ranges(
+            crai_index,
+            cram_reader,
+            |_| true,
+        )
+    }
+
+    /// Returns only unplaced unmapped ranges
+    fn get_byte_ranges_for_unmapped_reads(
+        &self,
+        crai_index: &[crai::Record],
+        cram_reader: &mut Reader<BufReader<File>>,
+    ) -> Result<Vec<BytesRange>> {
+        Self::get_bytes_ranges(
+            crai_index,
+            cram_reader,
+            |record| {
+                record.reference_sequence_id().map_or_else(
+                    || true,
+                    |ref_seq| i32::from(ref_seq) == reference_sequence_id::UNMAPPED,
+                )
+            },
+        )
+    }
+
+    /// Returns reads for a given reference name and an optional sequence range
+    fn get_byte_ranges_for_reference_name(
+        &self,
+        reference_name: &str,
+        crai_index: &[crai::Record],
+        cram_reader: &mut Reader<BufReader<File>>,
+        cram_header: Header,
+    ) -> Result<Vec<BytesRange>> {
+        let maybe_cram_ref_seq = cram_header.reference_sequences().get_full(reference_name);
+
+        let byte_ranges = match maybe_cram_ref_seq {
+            None => Err(HtsGetError::not_found(format!(
+                "Reference name not found: {}",
+                reference_name
+            ))),
+            Some((ref_seq_id, _, _)) => {
+                let cram_ref_seq_idx = ReferenceSequenceId::try_from(ref_seq_id as i32)
+                    .map_err(|_| HtsGetError::invalid_input("Invalid reference sequence id"))?;
+                Self::get_bytes_ranges(
+                    crai_index,
+                    cram_reader,
+                    |record| record.reference_sequence_id() == Some(cram_ref_seq_idx),
+                )
+            }
+        }?;
+        Ok(byte_ranges)
+    }
+
+    /// Get bytes ranges using the index.
+    fn get_bytes_ranges<F>(
+        crai_index: &[crai::Record],
+        cram_reader: &mut noodles_cram::Reader<BufReader<File>>,
+        predicate: F,
+    ) -> Result<Vec<BytesRange>>
+        where F: Fn(&Record) -> bool
+    {
+        // This could be improved by using some sort of index mapping.
+        let mut byte_ranges: Vec<BytesRange> = crai_index.iter().zip(crai_index.iter().skip(1))
+            .filter_map(|(record, next)| {
+                if predicate(record) {
+                    Some(BytesRange::default().with_start(record.offset()).with_end(next.offset()))
+                } else {
+                    None
+                }
+            }).collect();
+
+        let last = crai_index.last().ok_or_else(|| HtsGetError::invalid_input("No entries in CRAI"))?;
+        if predicate(last) {
+            // This operation is quite slow. An implementation based on file size might be better.
+            cram_reader.records().last();
```

Comment:
You should be able to use `cram_reader.seek` and the call `records`, instead of reading them all. 
The only problem I can see is that the `records` documentation states `The stream is expected to be at the start of a data container`. I don't know how difficult getting the byte offset of the start of the last container would be, but maybe this could be a solution. Of course, it's possible that using the file size, as you said, is easier 👍🏻

Reply:
That's a good point! I think that would work, and it would definitely make it quicker